architecture/adr-list/adr-008-estimating-transaction-fees.md

```diff
@@ -0,0 +1,253 @@
+   3. Exponential moving average
+3. Possible ways of implementation:
+   1. Client side
+   2. Provided as a service
+   3. On chain (the most trusted and robust way)
+3. To prevent constant growth of the mean value it's proposed to use multiplier `Y < 1` to get star fee price.
+4. If the dynamically calculated fees are insufficient to put a transaction though, use exponential growth to calculate higher values and retry.
+
+### Proposed transaction sending flow in VDR Tools
+
+1. Build and sign a transaction
+2. Send gas estimation request
+   1. Multiply response to the precalculated coefficient `X > 1` to increase probability of the transaction being committed
+3. Set initial gas price
+   1. Request moving average from the network
+   2. Multiply response by `Y < 1` to prevent constant growth of the mean value
+4. Try sending the transaction to the network
+   1. Exponentially increase the `gas` limit, in case of failure due to gas being lower than gas-wanted.
+   2. Exponentially increase the `gas-prices`, in case transaction time-out \(as this indicates `min-gas-prices` was not met\)
+
+### Proposed API in VDR Tools
+
+#### Build simulate query
+
+Used to build gas estimation request. 
+
+```
+indy_cheqd_ledger_tx_build_query_simulate(tx: byte[]) -> String
+```
+
+Parameters:
+* `tx` - transaction bytes from `build_tx` method
+
+Result:
+* request body
+
+`indy_cheqd_pool_abci_query` should be used to send the request and receive a response.
+
+#### Parse simulate query response
+
+Used to parse gas estimation response.
+
+```
+indy_cheqd_ledger_tx_parse_query_simulate_resp(resp: String) -> String
+```
+
+Parameters:
+* `resp` - response string from `indy_cheqd_pool_abci_query` call
+
+Result is json encoded  structure:
+
+```
+SimulateResponse {
+   GasInfo {
+      GasWanted uint64,
+      GasUsed uint64
+   },
+   Result {
+      ...
+   }
+}
+```
+
+Full type description can be found [here](https://github.com/cosmos/cosmos-sdk/blob/master/proto/cosmos/tx/v1beta1/service.proto#L112).
+
+`GasUsed` is the estimated gas value.
+
+#### Get gas multiplier
+
+It's recommended to use this precomputed multiplier to increase chances of a transaction being committed.
+
+```
+indy_cheqd_ledger_tx_get_gas_multiplier() -> double
+```
+
+Params:
+* None
+
+Result:
+* Multiplier value
+
+#### Build average gas price query
+
+```
+indy_cheqd_ledger_tx_build_query_prices() -> String
+```
+
+Parameters:
+* None
+
+Result:
+* Request body
+
+`indy_cheqd_pool_abci_query` should be used to send the request and receive a response.
+
+#### Parse average gas price response
+
+Used to parse gas estimation response.
+
+```
+indy_cheqd_ledger_tx_parse_query_prices_resp(resp: String) -> String
+```
+Params:
+* None
+
+Return string is json encoded structure:
+
+```
+PricesResponse {
+   LastAverage uint64
+   SubsetSize uint64
+}
+```
+
+`LastAverage` - average gas price for the recent `N` committed transactions where `N` is `SubsetSize`.
```

Comment:
A very large or very small value will move this figure too much potentially out of range

Reply:
The strategy should be chosen by a particular app/its back-end.